File: backend/app/registry/interpolate.py
"""Centripetal Catmull-Rom 1D interpolation — byte-parity with
`shared/registry/lib/interpolate-1d.ts`. Used by compound-widget ops.
"""
from __future__ import annotations

from typing import Any


def _pos(a: Any) -> float:
    return a["position"] if isinstance(a, dict) else a.position


def _vals(a: Any) -> dict[str, float]:
    return a["values"] if isinstance(a, dict) else a.values


def _catmull_rom(v0: float, v1: float, v2: float, v3: float, u: float) -> float:
    u2 = u * u
    u3 = u2 * u
    return 0.5 * (
        2 * v1
        + (-v0 + v2) * u
        + (2 * v0 - 5 * v1 + 4 * v2 - v3) * u2
        + (-v0 + 3 * v1 - 3 * v2 + v3) * u3
    )


def interpolate_1d(anchors: list[Any], t: float) -> dict[str, float]:
    """Interpolate derived values at position `t` along an anchor table.

    `anchors` is a list of dicts (or Pydantic models) each with `position`
    (float), `name` (str), and `values` (dict[str, float]). Must be sorted
    by position. Out-of-range `t` clamps to the nearest endpoint's values
    verbatim. Missing keys on a neighbour default to 0.
    """
    if len(anchors) < 2:
        raise ValueError("need at least 2 anchors")

    if t <= _pos(anchors[0]):
        return dict(_vals(anchors[0]))
    if t >= _pos(anchors[-1]):
        return dict(_vals(anchors[-1]))

    i = 0
    while i < len(anchors) - 1 and _pos(anchors[i + 1]) < t:
        i += 1
    p0 = anchors[max(i - 1, 0)]
    p1 = anchors[i]
    p2 = anchors[i + 1]
    p3 = anchors[min(i + 2, len(anchors) - 1)]

    span = _pos(p2) - _pos(p1)
    u = (t - _pos(p1)) / span if span > 0 else 0.0

    v0, v1, v2, v3 = _vals(p0), _vals(p1), _vals(p2), _vals(p3)
    keys = set(v0.keys()) | set(v1.keys()) | set(v2.keys()) | set(v3.keys())
    out: dict[str, float] = {}
    for k in keys:
        out[k] = _catmull_rom(
            v0.get(k, 0.0), v1.get(k, 0.0), v2.get(k, 0.0), v3.get(k, 0.0), u
        )
    return out


def interpolate_extended(anchors: list[Any], t: float) -> dict[str, float]:
    """`interpolate_1d`, plus linear extrapolation past the LAST anchor.

    Used by fused intent widgets whose driver overshoots the proposal
    (t in (1.0, 1.5]): the value continues along the last segment's slope.
    Below the first anchor it clamps exactly like `interpolate_1d`.
    Per-param range clamping is the CALLER's job (the registry knows ranges,
    this module doesn't).
    """
    if len(anchors) < 2:
        raise ValueError("need at least 2 anchors")
    last_pos = _pos(anchors[-1])
    if t <= last_pos:
        return interpolate_1d(anchors, t)

    prev, last = anchors[-2], anchors[-1]
    span = last_pos - _pos(prev)
    if span <= 0:
        return dict(_vals(last))
    pv, lv = _vals(prev), _vals(last)
    keys = set(pv.keys()) | set(lv.keys())
    overshoot = t - last_pos
    return {
        k: lv.get(k, 0.0) + ((lv.get(k, 0.0) - pv.get(k, 0.0)) / span) * overshoot
        for k in keys
    }

File: src/lib/perceptual-dial/interpolate.ts
import type { Anchor, CompoundParams } from './types';

/**
 * 1-D Catmull-Rom interpolation across `anchors` at scalar `t` in [0, 1].
 * Anchors are sorted internally by `position[0]`; ties keep first-seen order.
 * If `t` falls outside the anchor range, returns the nearest endpoint's params verbatim.
 * Missing keys on either neighbour default to 0 (so partial anchors interpolate towards 0).
 */
export function interpolate1D(anchors: Anchor[], t: number): CompoundParams {
  if (anchors.length === 0) return {};
  const sorted = [...anchors].sort((a, b) => a.position[0] - b.position[0]);
  if (t <= sorted[0].position[0]) return { ...sorted[0].params };
  if (t >= sorted[sorted.length - 1].position[0]) return { ...sorted[sorted.length - 1].params };

  // Find the segment [p1, p2] containing t.
  let i = 0;
  while (i < sorted.length - 1 && sorted[i + 1].position[0] < t) i += 1;
  const p0 = sorted[Math.max(i - 1, 0)];
  const p1 = sorted[i];
  const p2 = sorted[i + 1];
  const p3 = sorted[Math.min(i + 2, sorted.length - 1)];

  const span = p2.position[0] - p1.position[0];
  const u = span > 0 ? (t - p1.position[0]) / span : 0;

  // Collect the union of keys present across the four control anchors.
  const keys = new Set<string>([
    ...Object.keys(p0.params),
    ...Object.keys(p1.params),
    ...Object.keys(p2.params),
    ...Object.keys(p3.params),
  ]);

  const out: CompoundParams = {};
  for (const k of keys) {
    const v0 = p0.params[k] ?? 0;
    const v1 = p1.params[k] ?? 0;
    const v2 = p2.params[k] ?? 0;
    const v3 = p3.params[k] ?? 0;
    out[k] = catmullRom(v0, v1, v2, v3, u);
  }
  return out;
}

/**
 * Piecewise-linear interpolation across `anchors` at scalar `t` — mirrors
 * backend `interpolate_linear_1d`. Out-of-range `t` clamps to nearest endpoint.
 * Missing keys on a neighbour default to 0.
 */
export function interpolateLinear1D(anchors: Anchor[], t: number): CompoundParams {
  if (anchors.length === 0) return {};
  const sorted = [...anchors].sort((a, b) => a.position[0] - b.position[0]);
  if (t <= sorted[0].position[0]) return { ...sorted[0].params };
  if (t >= sorted[sorted.length - 1].position[0]) return { ...sorted[sorted.length - 1].params };

  let i = 0;
  while (i < sorted.length - 1 && sorted[i + 1].position[0] < t) i += 1;
  const p1 = sorted[i];
  const p2 = sorted[i + 1];

  const span = p2.position[0] - p1.position[0];
  const u = span > 0 ? (t - p1.position[0]) / span : 0;

  const keys = new Set<string>([...Object.keys(p1.params), ...Object.keys(p2.params)]);
  const out: CompoundParams = {};
  for (const k of keys) {
    const v1 = p1.params[k] ?? 0;
    const v2 = p2.params[k] ?? 0;
    out[k] = v1 + u * (v2 - v1);
  }
  return out;
}

/**
 * `interpolate1D` / `interpolateLinear1D` plus linear extrapolation past the
 * LAST anchor — mirrors backend `interpolate_extended`.
 *
 * `mode` selects in-range interpolation:
 * - `'catmull_rom_1d'` (default) — Catmull-Rom (back-compat for 2-anchor tables)
 * - `'linear_1d'` — piecewise-linear (3-anchor fused compounds)
 *
 * Extrapolation past the last anchor is always linear. Per-param range clamping
 * is the caller's job.
 */
export function interpolateExtended(
  anchors: Anchor[],
  t: number,
  mode: 'catmull_rom_1d' | 'linear_1d' = 'catmull_rom_1d',
): CompoundParams {
  if (anchors.length < 2) return interpolate1D(anchors, t);
  const sorted = [...anchors].sort((a, b) => a.position[0] - b.position[0]);
  const last = sorted[sorted.length - 1];
  if (t <= last.position[0]) {
    return mode === 'linear_1d' ? interpolateLinear1D(anchors, t) : interpolate1D(anchors, t);
  }

  const prev = sorted[sorted.length - 2];
  const span = last.position[0] - prev.position[0];
  if (span <= 0) return { ...last.params };
  const keys = new Set<string>([...Object.keys(prev.params), ...Object.keys(last.params)]);
  const overshoot = t - last.position[0];
  const out: CompoundParams = {};
  for (const k of keys) {
    const lv = last.params[k] ?? 0;
    const pv = prev.params[k] ?? 0;
    out[k] = lv + ((lv - pv) / span) * overshoot;
  }
  return out;
}

/** Centripetal-style Catmull-Rom scalar interpolation; tension 0.5 (standard). */
function catmullRom(v0: number, v1: number, v2: number, v3: number, u: number): number {
  const u2 = u * u;
  const u3 = u2 * u;
  return 0.5 * (
    (2 * v1) +
    (-v0 + v2) * u +
    (2 * v0 - 5 * v1 + 4 * v2 - v3) * u2 +
    (-v0 + 3 * v1 - 3 * v2 + v3) * u3
  );
}
